### PR TITLE
Docker user shell

### PR DESCRIPTION
**Pull request by wrinkledeth**
Added docker/user-shell folder which contains tooling to develop and build the user-shell container.

### CODE REVIEW

File: docker/user-shell/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+# Force cargo to fetch with git in order to avoid memory issues when running cargo build
+ENV CARGO_NET_GIT_FETCH_WITH_CLI=true
+
+# Clone and build Randcast Node Client
+RUN git clone https://github.com/ARPA-Network/BLS-TSS-Network.git && \
+    cd BLS-TSS-Network && \
+    cargo build --release
+
+####### Final Stage ########
+FROM rust:latest
```
Comment (mashuai) [inline docker/user-shell/Dockerfile:20]: don't need use rust base image, we can use debian-*-slim intead.

File: docker/user-shell/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+EXPOSE 8545 
+
+# Create a directory for the external config.yml file
+RUN mkdir -p /usr/src/app/external
+
+# Copy files from builder stage
+COPY --from=builder /usr/src/app/BLS-TSS-Network/target/release/user-shell /usr/src/app/user-shell
+
+# Install dependencies
+RUN apt-get update && \
```
Comment (mashuai) [inline docker/user-shell/Dockerfile:32]: As we don't need to compile user-shell, do we need to install the dependencies?

File: docker/user-shell/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,39 @@
+####### Builder Stage ########
+FROM rust:latest as builder
+
+# Define working directory
+WORKDIR /usr/src/app
+
+# Install dependencies
+RUN apt-get update && \
+    apt-get install -y protobuf-compiler libprotobuf-dev pkg-config libssh-dev build-essential supervisor neovim git net-tools netcat
```
Comment (mashuai) [inline docker/user-shell/Dockerfile:9]: Why we install neovim? And looks like that there is no need for supervisor.

File: docker/user-shell/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+# Install dependencies
+RUN apt-get update && \
+    apt-get install -y protobuf-compiler libprotobuf-dev pkg-config libssh-dev build-essential neovim git net-tools netcat
+
+# Install Foundry
+RUN curl -L https://foundry.paradigm.xyz | bash && \
+    /root/.foundry/bin/foundryup
+
+ENTRYPOINT ["/bin/bash", "-c"]
```
Comment (mashuai) [inline docker/user-shell/Dockerfile:39]: if you want to run interactive mode, you can use [interative](https://docs.docker.com/engine/reference/commandline/run/#examples) mode like `docker run --name test -it debian`.

File: docker/user-shell/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,39 @@
+####### Builder Stage ########
+FROM rust:latest as builder
```
Comment (mashuai) [inline docker/user-shell/Dockerfile:2]: how about we create our own base to build our project?

File: docker/user-shell/Dockerfile (hunk 4 above)
Comment (mashuai) [reply docker/user-shell/Dockerfile:39]: for this, I think it's ok, because we want to run the user-shell.

File: docker/user-shell/user_config.yml
Comment (kafeikui) [inline docker/user-shell/user_config.yml:1]: Let's remove this. As we are not going to copy a config to the container, the example in crates/user-cli will be enough.

**Review by kafeikui — APPROVED** [review]
120M is good to go for publishing the tool. Nice work and let's improve the steps later.

**Pull request merged**